public/setup/atomicupdate/alter_tables.php

```diff
@@ -0,0 +1,69 @@
+<?php
+
+$do_after_services = true;
+
+$personnel = array();
+
+$db = new db();
+$db->select2('personnel', array('id', 'nom', 'prenom', 'service'));
+
+
+if($db->result) {
+    foreach($db->result as $elem) {
+        $personnel[$elem['id']] = array("id" => $elem['id'], "nom" => $elem['nom'], "prenom" => $elem['prenom'], "service" => $elem['service']);
+    }
+}
+
+$db = new db();
+$db->select2("select_services",array("id","valeur"), null,null);
+$services = $db->result;
+if($services){
+    foreach($services as $elem){
+        $sql[] = "UPDATE `{$dbprefix}personnel` SET `service`= '{$elem['id']}'  WHERE `service` = '{$elem['valeur']}';";
+    }
+}
+
+
+$sql[] = "ALTER TABLE `{$dbprefix}personnel` MODIFY `service` int(11) NOT NULL DEFAULT 0;";
+
+$db2 = new db();
+$db2->select("select_services","*", null,null);
+if($db2->result){
+    foreach ($db2->result as $elem) {
+        $id = $elem['id'];
+        $old = $elem['valeur'];
+        $new = html_entity_decode($old, ENT_QUOTES|ENT_IGNORE, 'UTF-8');
+        if ($new != $old) {
+            $new = addslashes($new);
+            $sql[] = "UPDATE `{$dbprefix}select_services` SET `valeur`= '$new'  WHERE `id` = '$id';";
+        }
+    }
+}
+
+$after[] = function() {
+    global $do_after_services, $personnel;
+    $cpt = 0;
+    if(isset($do_after_services)){
+        $db = new db();
+        $db->select2('personnel', array('id', 'nom', 'prenom', 'service'), array("service"=> "= 0"));
+
+        if ($db->result){
+            foreach($db->result as $agent) {
+                $id = $agent['id'];
+                if ($personnel[$id]['service'] != ''){
+                    $cpt++;
+                }
+            }
+            if ($cpt > 0){
+                echo  "Le service des agents suivants doit être vérifié :<br>\n";
+                foreach($db->result as $agent) {
+                    $id = $agent['id'];
+                    if (empty($personnel[$id]['service'])) {
+                        continue;
```

Comment:
Je viens de penser à quelque chose de plus simple encore, et qui t'éviterait de compter.
En ligne 8, 
<pre>db->select2('personnel', array('id', 'nom', 'prenom', 'service'));</pre>
devient
<pre>db->select2('personnel', array('id', 'nom', 'prenom', 'service'), "WHERE  ̀service` <> ''");</pre>

A part la ligne 8, on reviendrait au code précédent. Le traitement sera plus rapide, même s'il convient dans l'état actuel.
A toi de voir.

Reply:
C'est intégré. Reste à supprimer la première ligne d'affichage ```Le service des agents suivants doit être vérifié :<br> ```

Reply:
Logiquement, la ligne ne s'affichera pas si elle n'est pas nécessaire car elle est dans le if($db->result), qui devrait être vide si aucune erreur, car le filtre service <> '' existe toujours, il est simplement déplacé plus haut. Je teste et je te dis.

Reply:
Ben si, il affiche la ligne )-;
C'est parce que le contrôle 
<pre>if (in_array($id, $personnel) and $personnel[$id]['service'] == ''){</pre>
arrive après (tu as bien fait de la mettre).
tu peux remettre le comptage sur ce nouveau code (désolé de t'avoir incité à l'enlever).

Reply:
Je suis en train de me rendre compte que c'est != et pas ==.... Je vais mettre à jour tout ça...